#!/bin/bash

#set -e

#########################################################
#
# Platform: NCI Gadi HPC
#
# Author: Tracy Chew
# tracy.chew@sydney.edu.au
#
# If you use this script towards a publication, please acknowledge the
# Sydney Informatics Hub (or co-authorship, where appropriate).
#
# Suggested acknowledgement:
# The authors acknowledge the scientific and technical assistance
# <or e.g. bioinformatics assistance of <PERSON>> of Sydney Informatics
# Hub and resources and services from the National Computational
# Infrastructure (NCI), which is supported by the Australian Government
# with access facilitated by the University of Sydney.
#
#########################################################

# Description: Performs adapter and polyA tail trimming 
# Runs bbduk_trim_paired.sh in parallel if bbduk_trim_paired.inputs exists.
# Runs bbduk_trim_single.sh in parallel if bbduk_trim_single.inputs exists.
# Will automatically look for either input file.
# For 15 FASTQ pairs, ~80 M reads each, use
# -l walltime=02:00:00,ncpus=48,mem=190GB,wd
# -q normal

#PBS -P er01
#PBS -N bbduk_trim
#PBS -l walltime=02:00:00,ncpus=48,mem=190GB,wd
#PBS -q normal
#PBS -W umask=022
#PBS -l storage=scratch/er01
#PBS -o ./Logs/bbduk_trim.o
#PBS -e ./Logs/bbduk_trim.e

module load openmpi/4.1.0
module load nci-parallel/1.0.0a
#module load bbtools/37.98

mkdir -p ./Logs

export PATH=$PATH:/scratch/er01/INFRA-121-CoreWorkflows-RNASeq/bbmap


# NCPUs = CPUs per task
NCPUS=6

INPUTS=./Inputs
paired=${INPUTS}/bbduk_trim_paired.inputs
single=${INPUTS}/bbduk_trim_single.inputs

#########################################################
# Do not edit below this line
#########################################################

M=$(( PBS_NCI_NCPUS_PER_NODE / NCPUS )) #tasks per node

if [ -s ${paired} ]
then
	echo "$(date): Found inputs file for paired data"
	SCRIPT=./bbduk_trim_paired.sh
	sed "s|^|${SCRIPT} |" ${paired} > ${PBS_JOBFS}/input-file-paired
	mpirun --np $(( M * PBS_NCPUS/PBS_NCI_NCPUS_PER_NODE )) \
		--map-by node:pe=${NCPUS} \
		--verbose \
		nci-parallel \
		--input-file ${PBS_JOBFS}/input-file-paired
fi

if [ -s ${single} ]
then
	echo "$(date): Found inputs file for single data"
	SCRIPT=./bbduk_trim_single.sh
	sed "s|^|${SCRIPT} |" ${single} > ${PBS_JOBFS}/input-file-single
	mpirun --np $(( M * PBS_NCPUS/PBS_NCI_NCPUS_PER_NODE )) \
		--map-by node:pe=${NCPUS} \
		--verbose \
		nci-parallel \
		--input-file ${PBS_JOBFS}/input-file-single
fi
