#!/bin/bash

#set -e

#########################################################
#
# Platform: NCI Gadi HPC
#
# Author: Tracy Chew
# tracy.chew@sydney.edu.au
#
# If you use this script towards a publication, please acknowledge the
# Sydney Informatics Hub (or co-authorship, where appropriate).
#
# Suggested acknowledgement:
# The authors acknowledge the scientific and technical assistance
# <or e.g. bioinformatics assistance of <PERSON>> of Sydney Informatics
# Hub and resources and services from the National Computational
# Infrastructure (NCI), which is supported by the Australian Government
# with access facilitated by the University of Sydney.
#
#########################################################

# can map paired and single reads in a single job
# Should scale to the number of reads in FASTQ file

#======================================================================================
#                  Resource Usage on 2023-05-12 22:01:30:
#   Job Id:             83363636.gadi-pbs
#   Project:            er01
#   Exit Status:        0
#   Service Units:      8.59
#   NCPUs Requested:    12                     NCPUs Used: 12
#                                        CPU Time Used: 02:39:10
#   Memory Requested:   64.0GB                Memory Used: 35.93GB
#   Walltime requested: 02:00:00            Walltime Used: 00:16:06
#   JobFS requested:    100.0MB                JobFS used: 0B
#======================================================================================



#PBS -P er01
#PBS -N salmon_align_trimmed
#PBS -l walltime=01:00:00,ncpus=96,mem=240GB,wd
#PBS -q normal
#PBS -W umask=022
#PBS -l storage=scratch/er01+gdata/er01
#PBS -o ./Logs/salmon_align_trimmed_unmapped_out.o
#PBS -e ./Logs/salmon_align_trimmed_unmapped_out.e

module load openmpi/4.0.2
module load nci-parallel/1.0.0a
#Nandan

mkdir -p ./Logs

# NCPUs = CPUs per task
NCPUS=12

INPUTS=./Inputs
paired=${INPUTS}/salmon_align_paired.inputs
single=${INPUTS}/salmon_align_single.inputs

PAIRED_SCRIPT=./salmon_align_paired.sh
SINGLE_SCRIPT=./salmon_align_single.sh




# Activate the conda environment
conda init bash
source ~/.bashrc
conda activate salmon


#########################################################
# Do not edit below this line
#########################################################

# Number of tasks per node
M=$(( PBS_NCI_NCPUS_PER_NODE / NCPUS ))

if [ -s ${paired} ]
then
	echo "$(date): Found inputs file for paired data. Performing alignment with STAR for paired data."
	SCRIPT=${PAIRED_SCRIPT}
	sed "s|^|${SCRIPT} |" ${paired} > ${PBS_JOBFS}/input-file-paired
	mpirun --np $((M * PBS_NCPUS / PBS_NCI_NCPUS_PER_NODE)) \
		--map-by node:PE=${NCPUS} \
		nci-parallel \
		--verbose \
		--input-file ${PBS_JOBFS}/input-file-paired
fi

if [ -s ${single} ]
then
	echo "$(date): Found inputs file for single data. Performing alignment with STAR for single read data."
	SCRIPT=${SINGLE_SCRIPT}
	sed "s|^|${SCRIPT} |" ${single} > ${PBS_JOBFS}/input-file-single
        mpirun --np $((M * PBS_NCPUS / PBS_NCI_NCPUS_PER_NODE)) \
                --map-by node:PE=${NCPUS} \
                nci-parallel \
                --verbose \
                --input-file ${PBS_JOBFS}/input-file-single
fi
